src/models/files/file-folder.js

```
@@ -0,0 +1,170 @@
+        }
+        this.files.push(file);
+    }
+
+    addFolder(folder) {
+        if (folder.parent === this) return folder;
+        if (this.folderMap[folder.folderId]) {
+            console.error('folder already exists here');
+            return folder;
+        }
+        if (folder.parent) {
+            console.debug('moving folder from parent');
+            folder.parent.freeFolder(folder);
+        }
+        folder.parent = this;
+        this.folders.push(folder);
+        return folder;
+    }
+
+    free(file) {
+        if (!this.fileMap[file.fileId]) {
+            console.error('file does not belong to the folder');
+            return;
+        }
+        const i = this.files.indexOf(file);
+        if (i !== -1) {
+            this.files.splice(i, 1);
+            file.folder = null;
+        } else {
+            console.error('free cannot find the file');
+        }
+    }
+
+    freeFolder(folder) {
+        const i = this.folders.indexOf(folder);
+        if (i !== -1) {
+            this.folders.splice(i, 1);
+            folder.parent = null;
+        } else {
+            console.error('free cannot find the folder');
+        }
+    }
+
+    freeSelf() {
+        if (this.isRoot) return;
+        let root = this;
+        while (!root.isRoot) root = root.parent;
+        this.files.forEach(file => {
+            file.folder = null;
+            root.add(file);
+        });
+        this.files = [];
+        this.folders.forEach(folder => folder.freeSelf());
+        this.folders = [];
+        this.parent && this.parent.freeFolder(this);
+    }
+
+    moveInto(file) {
+        if (file.isFolder) {
+            if (this.findFolderByName(file.normalizedName)) {
+                warnings.addSevere('error_folderAlreadyExists');
+                throw new Error('error_folderAlreadyExists');
+            }
+            if (file === this) {
+                console.error('cannot move folder in itself');
+                return;
```

See what i mean with previous comment? Here we just silently return. Not even a warning. Need consistency.